Adding UPS coordinates functionality to keep parity with USNG.js

dimitry-sherman opened the pull request:
#### What does this PR do?
Adding UPS coordinates functionality to keep parity with USNG.js

#### Who is reviewing it? 
@millerw8
@Lambeaux
@Variadicism
@samuelechu

#### Select relevant component teams: 
https://github.com/orgs/codice/teams
#### Choose 2 committers to review/merge the PR. 
(please choose ONLY two committers from below, delete the rest)
@stustison
@dcruver

#### How should this be tested? (List steps with links to updated documentation)
All the necessary unit tests have been added and pass, all the legacy unit tests pass as well.
**Modified code code coverage:**
CoordinateSystemTranslatorImpl  |  Method - 80% (28/ 35)  |  Line - 83.6% (310/ 371)
**Added code code coverage:**
UtmUpsCoordinateImpl  |  Method - 100% (28/ 28)  |  Line - 100% (94/ 94)

#### Any background context you want to provide?
Added UPS functionality to `usng4j` library for parity wiht `usng.js` where the functionality was already added.
#### What are the relevant tickets?
[DDF-4323](https://codice.atlassian.net/browse/DDF-4323)
#### Screenshots (if appropriate)
#### Checklist:
- [x] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

Commit:
Added initial skeleton for the UtmUpsCoordinateImpl unit tests.

samuelechu committed:
Continue unit tests skeleton

samuelechu committed:
Fix package folder structure

Variadicism committed:
Rearranged and tweaked test definintions

I rearranged the test methods to group them and make the test case enumeration patterns easier to see. I grouped them by UTM, UPS, and UTM/UPS and arranged the tests within those groups as valid cases, then invalid cases, then overlapping cases.

I also removed some terms in test method names that seemed to be unnecessary; the logic is described below:
- Specifying that a UTM string is "valid" is redundant; if the string is not valid, then it is not UTM at all.
- Specifying that we are parsing strings seems unnecessary; parsing of coordinates will certainly take place on some form of text and all parsing methods that we plan to implement take strings only, so there is nothing to differentiate it from.
- Tests with specifications like "testParsingInvalidUtmAsUtm" can be simplified to "testParsingInvalidUtm"; as stated before, invalid UTM is not truly UTM at all, so saying that you're parsing invalid UTM can only imply that you're parsing it as UTM, making "AsUtm" redundant.

Commit:
Merge pull request #1 from Variadicism/addingUpsCoordinates

Rearranged and tweaked test definintions

samuelechu committed:
Add UtmUpsCoordinateImpl and gson dependency

Variadicism committed:
Added CoordinatePrecision and NSIndicator fields and associated uses

Commit:
Updated `UtmUpsCoordinateImplTest` to read test objects into their own class instances.

Commit:
Formatting fix.

Commit:
Extracted common test functionality to avoid code duplication.

Commit:
Added `String` representations to the test data `.json` file, and made necessary modifications to the `UtmUpsTestData` class.

Variadicism committed:
[WIP] Adding UTMUPS to Lat/Lon conversion

Commit:
Refactored changes made bf @Variadicism to pass the quick-build (using `cx-build`), and added some backward compatibility code.
Additionally removed unnecessary for JDK8 JavaDoc comment.

Commit:
[WIP - buildable] Added first draft of input validation, please check and correct/add what's wrong or missing.

Commit:
[WIP - buildable] Added the interface changes to `CoordinateSystemTranslator` for the `toUps` and `toUtmUps` API.

Variadicism committed:
Complete implementation of UPS-LL translation and addition of UPS and UTM/UPS to LL methods in CoordinateSystemTranslator

All tests pass currently, but more unit tests need to be added to test the new functionality.

Commit:
[WIP - buildable] Added valid UTM and UPS bands and min/max unified range for Easting and Northing to use in validation, and later for `isUTM` and `isUPS`.

Commit:
Fixed NORTHEN_OFFSET based on NSIndicator inconsistencies, removed some not needed in JDK8 comments.  All existing tests run, the changes are backwards compatible.

Commit:
Removed the unnecessary NPE check.

Commit:
Added `@Deprecated` to the misspelled method and the correctly spelled implementation, also removed unneeded in JDK8 comments.

samuelechu committed:
Add utm/ups string parsing function + some tests

Commit:
Auto stash before rebase of "forked/addingUpsCoordinates"

Commit:
Refactored UTM/UPS parsing code a little for readability and consistency.

Commit:
Added a few more tests needing implementation.

Commit:
Fixed the bug found by failing `testParseUtmBadInput`.

Commit:
Formatting plug-in reformatted some code.

Commit:
Added test data iteration code by @Variadicism (slightly generalized for easier re-use).

Commit:
Added test(s) for LLtoUPS (or more accurately DecimalDegrees `toUPS`), additionally refactored/corrected some of the testing harness code.

Commit:
[The two files missed in the previous commit] Added test(s) for LLtoUPS (or more accurately DecimalDegrees `toUPS`), additionally refactored/corrected some of the testing harness code.

Commit:
Implemented `toUps`.

Commit:
Some minor clean-up of warnings.

Variadicism committed:
Completed UTM/UPS to Lat-Lon unit tests

Doing this required some fixes to the UtmUpsCoordinate factory methods, fixing errors in the UPS test data, and debugging math.

Commit:
[WIP - tests are 1 to 3 meters off] Added `toUPS` from DecimalDegree functionality.

Commit:
Completed `toUPS` from DecimalDegree functionality, commented out questionable test data.

Removed unused code.

Commit:
Finished all outstanding UT(s), updated POM file with new minimum coverages.

dimitry-sherman commented:
> It doesn't make sense to me to return an `Optional` version of the parameter unless the object passed in is potentially different from the object returned inside the `Optional`. I think that this method signature could be simplified to `private static bool isValid(UtmUpsCoordinate utmUpsCoordinate)`.

This is used in the additional **Factory Method** ending with `...IfPossible`.

Commit:
Forgot to commit these tests.

coyotesqrl commented on usng4j-api/src/main/java/org/codice/usng4j/UtmUpsCapabilities.java:
```diff
@@ -24,12 +24,19 @@
   int getZoneNumber();
 
   /** @return the latitude band for this UTM coordinate or null if not specified. */
+  @Deprecated
```
✏️ Nit: should include a link to the new, correctly spelled `{@link #getLatitudeBand}` method.

coyotesqrl commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/CoordinateSystemTranslatorImpl.java:
```diff
@@ -279,13 +291,49 @@ public UtmCoordinate toUtm(final DecimalDegreesCoordinate decimalDegreesCoordina
+  @Override
+  public UtmUpsCoordinate toUtmUps(final DecimalDegreesCoordinate decimalDegreesCoordinate) {
+    validateDecimalDegreeInput(decimalDegreesCoordinate);
+    final boolean northPole = decimalDegreesCoordinate.getLat() >= 0.0;
```
❓ This boolean represents that the latitude is in the northern hemisphere, not at the north pole, right? If so, should rename to clarify that. Maybe `isNorthern` or similar.

dimitry-sherman replied:
This used for calculating UPS and hence determines the poles.

coyotesqrl commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/CoordinateSystemTranslatorImpl.java:
```diff
@@ -552,24 +590,85 @@ BoundingBox toBoundingBox(final UtmCoordinate utmCoordinate, final Integer accur
     } else {
-      return toLatLonNsNormalized(utmCoordinate);
+      final double rho = Math.hypot(easting, northing);
+      final double t = rho != 0.0 ? rho / RHO_ADJUSTER_VALUE : Math.pow(EPSILON, 2.0);
```
✏️ I don't personally find ternaries with nested math processing especially easy to read. It is only a matter of preference, but when the get longer like this, I would much prefer they be expanded to if/else blocks.

dimitry-sherman replied:
This was ported as is from GeographicLib, first into USNG.js and then here; we didn't want to break the actual calculations as we don't know all the math for it.

coyotesqrl commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/CoordinateSystemTranslatorImpl.java:
```diff
@@ -552,24 +590,85 @@ BoundingBox toBoundingBox(final UtmCoordinate utmCoordinate, final Integer accur
+  private static double taupf(final double tauValue) {
+    final double tau1 = Math.hypot(1.0, tauValue);
+    final double sig = Math.sinh(eatanhe(tauValue / tau1));
+    return Math.hypot(1.0, sig) * tauValue - sig * tau1;
```
✏️ I know that I prefer explicitly declaring operation order with parens rather than needing to remember precedence rules.

coyotesqrl commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/CoordinateSystemTranslatorImpl.java:
```diff
@@ -552,24 +590,85 @@ BoundingBox toBoundingBox(final UtmCoordinate utmCoordinate, final Integer accur
+    // is used the mean number of iterations increases to 1.99 (2 iterations
+    // are needed except near tau = 0).
+    double tau = taupValue / e2m;
+    final double stol = Math.sqrt(EPSILON) / 10.0 * Math.max(1.0, Math.abs(taupValue));
```
✏️ I know that I prefer explicitly declaring operation order with parens rather than needing to remember precedence rules.

Variadicism committed:
Corrected UPS test data and increased easting/northing acceptable deviation to 4m

coyotesqrl commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/CoordinateSystemTranslatorImpl.java:
```diff
@@ -23,27 +23,36 @@
 
   private static final int BLOCK_SIZE = 100000;
 
+  private static final double EPSILON = Math.ulp(1.0);
```
❓ Why is this the correct value for `EPSILON` not not `Math.ulp(2.0)` or any other input float? Is it somewhat arbitrary, determined empirically, or dictated by some aspect of the spec?

Variadicism-zz replied:
This epsilon value is intended to be the smallest representable difference between 1.0 and the next double value to match the behavior of these calculations from the original C++ code that it was ported from (which uses [numeric_limits::epsilon](https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon)) and from usng.js (which uses [Number.EPSILON](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/EPSILON)).

coyotesqrl commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/UtmUpsCoordinateImpl.java:
```diff
@@ -0,0 +1,292 @@
+  private static final int NORTHING_RE_GROUP = 4;
+  private static final int NS_INDICATOR_RE_GROUP = 5;
+  private static final Set<Character> upsNorthenBands = new HashSet<>(Arrays.asList('Y', 'Z'));
+  private static final Set<Character> upsSothernBands = new HashSet<>(Arrays.asList('A', 'B'));
```
✏️ s/b `upsSouthernBands`

coyotesqrl commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/UtmUpsCoordinateImpl.java:
```diff
@@ -0,0 +1,292 @@
+  private static final int EASTING_RE_GROUP = 3;
+  private static final int NORTHING_RE_GROUP = 4;
+  private static final int NS_INDICATOR_RE_GROUP = 5;
+  private static final Set<Character> upsNorthenBands = new HashSet<>(Arrays.asList('Y', 'Z'));
```
✏️ s/b `upsNorthernBands`

coyotesqrl commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/UtmUpsCoordinateImpl.java:
```diff
@@ -0,0 +1,292 @@
+  private static final int NS_INDICATOR_RE_GROUP = 5;
+  private static final Set<Character> upsNorthenBands = new HashSet<>(Arrays.asList('Y', 'Z'));
+  private static final Set<Character> upsSothernBands = new HashSet<>(Arrays.asList('A', 'B'));
+  private static final Set<Character> utmNorthenBands =
```
✏️ s/b `utmNorthernBands`

Variadicism committed:
Quote JSON keys for true JSON compliance

The test data worked before because GSON allows keys to not be quoted, but standard JSON must quote all keys, so for greater compliance with more development tools and in support of the true standard, the keys are now all quoted. GSON operates on the data properly either way.

coyotesqrl commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/UtmUpsCoordinateImpl.java:
```diff
@@ -0,0 +1,292 @@
+  private static final Set<Character> upsSothernBands = new HashSet<>(Arrays.asList('A', 'B'));
+  private static final Set<Character> utmNorthenBands =
+      new HashSet<>(Arrays.asList('C', 'D', 'E', 'F', 'G', 'H', 'J', 'K', 'L', 'M'));
+  private static final Set<Character> utmSothernBands =
```
✏️ s/b `utmSouthernBands`

coyotesqrl commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/UtmUpsCoordinateImpl.java:
```diff
@@ -0,0 +1,292 @@
+  }
+
+  static boolean isUpsOverlapException(final UtmUpsCoordinate candidateCoordinate) {
+    // TODO:  implement overlap checking
```
❓ Would it be better to throw an `UnsupportedOperationException` for now instead of returning false?

Either way, is there a ticket to resolve this that could be referenced here?

dimitry-sherman replied:
The fix has been posted.

Variadicism committed:
Implement isUTM and isUPS fully.

This change asserts that it is more intuitive and useful to make such methods check whether the data contained in a `UtmUpsCoordinate` describe coordinates in UTM space or in UPS space regardless of whether or not that UTM coordinate overlaps UPS space or vice versa.

Variadicism committed:
Removed unfinished test for functionality that is no longer applicable

Variadicism committed:
Format

Variadicism committed:
Renamed factory method to match argument order

The factory method's name was intended to reflect the order and names of the arguments that it takes, but while the method name listed "Northing" before "Easting", the argument order put easting before northing. Usages of this method were correct, but the name was misleading for future usages.

Commit:
Updated code coverage minimums (increased).
